src/cargo/lib.rs

```diff
@@ -195,37 +199,49 @@ pub fn exit_with_error(err: CliError, shell: &mut MultiShell) -> ! {
     std::process::exit(exit_code)
 }
 
-pub fn handle_error(err: &CargoError, shell: &mut MultiShell) {
-    debug!("handle_error; err={:?}", err);
+pub fn handle_error(err: CargoError, shell: &mut MultiShell) {
+    debug!("handle_error; err={:?}", &err);
 
-    let _ignored_result = shell.error(err);
+    let _ignored_result = shell.error(&err);
     handle_cause(err, shell);
 }
 
-fn handle_cause(mut cargo_err: &CargoError, shell: &mut MultiShell) -> bool {
-    let verbose = shell.get_verbose();
-    let mut err;
-    loop {
-        cargo_err = match cargo_err.cargo_cause() {
-            Some(cause) => cause,
-            None => { err = cargo_err.cause(); break }
-        };
-        if verbose != Verbose && !cargo_err.is_human() { return false }
-        print(cargo_err.to_string(), shell);
-    }
-    loop {
-        let cause = match err { Some(err) => err, None => return true };
-        if verbose != Verbose { return false }
-        print(cause.to_string(), shell);
-        err = cause.cause();
-    }
-
+fn handle_cause<E, EKind>(cargo_err: E, shell: &mut MultiShell) -> bool where E: ChainedError<ErrorKind=EKind> + 'static {
     fn print(error: String, shell: &mut MultiShell) {
         let _ = shell.err().say("\nCaused by:", BLACK);
         let _ = shell.err().say(format!("  {}", error), BLACK);
     }
+
+    unsafe fn extend_lifetime(r: &Error) -> &'static Error {
```

Comment:
This could have a safer signature:

```
    unsafe fn extend_lifetime(r: &Error) -> &(Error + 'static) {
        std::mem::transmute::<&Error, &Error>(r)
    }
```

It would be nice to add some short comment here, explaining why we need unsafe.